AttributeError: '_ClientMethodFactory' object has no attribute '__code__'

Issue:
Running make docs throw a AttributeError.
```
(v3nv) quokka@qcore:~/git/delet/dcp-cli$ make docs
make -C docs html
make[1]: Entering directory '/home/quokka/git/delet/dcp-cli/docs'
Running Sphinx v2.1.2
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 3 source files that are out of date
updating environment: 3 added, 0 changed, 0 removed
reading sources... [ 33%] api                                                   
Exception occurred:
  File "/usr/lib/python3.6/inspect.py", line 186, in iscoroutinefunction
    object.__code__.co_flags & CO_COROUTINE)
AttributeError: '_ClientMethodFactory' object has no attribute '__code__'
The full traceback has been saved in /tmp/sphinx-err-6jwazei4.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Makefile:20: recipe for target 'html' failed
make[1]: *** [html] Error 2
make[1]: Leaving directory '/home/quokka/git/delet/dcp-cli/docs'
Makefile:47: recipe for target 'docs' failed
make: *** [docs] Error 2
```
Errors here:
https://github.com/HumanCellAtlas/dcp-cli/blob/34e6d610578c387726689861c7bb9bf62638e151/hca/util/__init__.py#L128

--- a/CHANGES
+++ b/CHANGES
@@ -56,6 +56,7 @@ Bugs fixed
   imports when ``'bysource'`` order
 * #6574: autodoc: missing type annotation for variadic and keyword parameters
 * #6589: autodoc: Formatting issues with autodoc_typehints='none'
+* #6605: autodoc: crashed when target code contains custom method-like objects
 * #6498: autosummary: crashed with wrong autosummary_generate setting
 * #6507: autosummary: crashes without no autosummary_generate setting
 * #6511: LaTeX: autonumbered list can not be customized in LaTeX

--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -207,9 +207,12 @@ def isbuiltin(obj: Any) -> bool:
 
 def iscoroutinefunction(obj: Any) -> bool:
     """Check if the object is coroutine-function."""
-    if inspect.iscoroutinefunction(obj):
+    if hasattr(obj, '__code__') and inspect.iscoroutinefunction(obj):
+        # check obj.__code__ because iscoroutinefunction() crashes for custom method-like
+        # objects (see https://github.com/sphinx-doc/sphinx/issues/6605)
         return True
-    elif ispartial(obj) and inspect.iscoroutinefunction(obj.func):
+    elif (ispartial(obj) and hasattr(obj.func, '__code__') and
+          inspect.iscoroutinefunction(obj.func)):
         # partialed
         return True
     else:

